### PR TITLE
Feature/card card datable implementaion

### PR DESCRIPTION
**Pull request by purushothamanTR**
datatable and regular expression unittest case release

### CODE REVIEW

File: BookingSystem.Tests/APIController.Tests/APICardController.Tests/GetCardDatatableTest.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            // Arrange
+            var responseData = fixture.Create<ResponseModel>();
+
+            var cardData = fixture.Build<CardDatatableVM>()
+                .Without(x => x.carddetails)
+                .Create();
+
+            var cardsVMData = fixture.Build<CardsVM>().With(a => a.IsDeleted, false).With(x => x.CustomerID, 1).With(x => x.IsDeleted, false)
+                .With(x => x.CVV_No, "124")
+                .With(x => x.CardNumber, "123")
```
Comment (manic1990) [inline BookingSystem.Tests/APIController.Tests/APICardController.Tests/GetCardDatatableTest.cs:66]: Use constant don't hard code value.

File: BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetCard.cs [hunk 2]
```diff
@@ -32,7 +32,12 @@ public async void GetCard_ShouldGetCardRecord()
         {
             //Arrange
             var responseData = fixture.Create<ResponseModel>();
-            var cardData = fixture.Create<CardsVM>();
+            var cardData = fixture.Build<CardsVM>().With(a => a.IsDeleted, false).With(x => x.CustomerID, 1).With(x => x.IsDeleted, false)
+.With(x => x.CVV_No, "124")
+.With(x => x.CardNumber, "123")
```
Comment (manic1990) [inline BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetCard.cs:37]: Use constant don't hard code value check it other places also.

File: BookingSystem/Controllers/CardController.cs [hunk 3]
```diff
@@ -165,9 +180,12 @@ public async Task<IActionResult> GetData(ViewDatatableParam JqueryDatatableParam
             catch (Exception ex)
             {
 
-                throw;
+                _logger.LogError(ex.Message);
             }
+
+
             _logger.LogInformation(nameof(CardController) + "." + nameof(Card) + complete);
+            return null;
```
Comment (manic1990) [inline BookingSystem/Controllers/CardController.cs:188]: Use return default(ViewCardDatatable);

File: BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetDatatableCard.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
+                .ReturnsAsync(new HttpResponseMessage
+                {
+                    StatusCode = HttpStatusCode.BadRequest,
+                    Content = null,
+                });
+
+            var httpClient = new HttpClient(mockMessageHandler.Object);
+            httpClient.BaseAddress = new Uri("https://localhost:44348/api/");
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetDatatableCard.cs:117]: Refer someother testcase, Keep uri value in appsetting.json

File: BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetDatatableCard.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                {
+                    StatusCode = HttpStatusCode.OK,
+                    Content = content,
+                });
+
+            var httpClient = new HttpClient(mockMessageHandler.Object);
+            httpClient.BaseAddress = new Uri("https://localhost:44348/api/");
+
+            mockFactory.Setup(a => a.CreateClient(It.IsAny<string>())).Returns(httpClient);
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/ApiClient.Tests/Card/APIClientGetDatatableCard.cs:77]: Keep url in appsetting json. refer some other API testcases.

File: BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                iSortCol_0 = 0,
+                sSearch = null
+            };
+
+
+            var exceptedCustomerId = card.CardID;
+            //create In Memory Database
+            var options = new DbContextOptionsBuilder<DataContext>().UseInMemoryDatabase(databaseName: "BookingSystem").Options;
+            //var context = new DataContext(options);
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs:79]: Refer someother methods, dont keep hard code values for database

File: BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardTest.cs [hunk 7]
```diff
@@ -45,18 +46,29 @@ public async Task GetCardDetails_ShouldGetCardRecord()
...
 
+
+            var customerData = fixture.Build<CustomerVM>().With(a => a.IsDeleted, false)
+    .With(x => x.CustomerID, 1).Create();
+            var customer = _testmapper.Map<Customer>(customerData);
+
             var exceptedCustomerId = card.CardID;
             //create In Memory Database
             var options = new DbContextOptionsBuilder<DataContext>().UseInMemoryDatabase(databaseName: "BookingSystem").Options;
             //var context = new DataContext(options);
```
Comment (Backiaraj07) [inline BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardTest.cs:65]: Dont give Hard code value for database name

File: BookingSystem/Controllers/CardController.cs [hunk 8]
```diff
@@ -155,8 +156,22 @@ public async Task<IActionResult> GetData(ViewDatatableParam JqueryDatatableParam
...
+                    sortDirection = HttpContext.Request.Query["sSortDir_0"];
+                }
+                else
+                {
+                    sortColumnIndex = 1;
+                    sortDirection = "asc";
+
+                }
+
                 var result = await _cardRepository.sortingAsync(JqueryDatatableParam, sortColumnIndex, sortDirection).ConfigureAwait(false);
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/CardController.cs:175]: write a log before of return, once return runs it wont go for next line

Comment (Backiaraj07) [reply BookingSystem/Controllers/CardController.cs:175]: change it to all methods

**Review by Backiaraj07 — COMMENTED** [review]
Added comments pls fix it

**Review by manic1990 — APPROVED** [review]

File: BookingSystem.Tests/Controller.Tests/CardControllers.Test/SaveCardTest.cs [hunk 9]
```diff
@@ -48,10 +48,13 @@ public SaveCardTest()
         public async void GetActiveCardAsync_ReturnsOkResult()
         {
             // Arrange
-            var data = fixture.Create<CardsVM>();
+  var data = fixture.Build<CardsVM>().With(a => a.IsDeleted, false).With(x => x.CustomerID, 1).With(x => x.IsDeleted, false)
+    .With(x => x.CVV_No, "124")
+    .With(x => x.CardNumber, "123")
+  .With(x => x.CardID, 1).Create();
```
Comment (Backiaraj07) [inline BookingSystem.Tests/Controller.Tests/CardControllers.Test/SaveCardTest.cs:54]: dont hardcode the values

File: BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        [Fact]
+        public async Task GetCardDetails_ShouldGetCardRecord()
+        {
+            //Arrange
+            var cardData = fixture.Build<CardsVM>().With(a => a.IsDeleted, false).With(x => x.CustomerID, 1).With(x => x.IsDeleted, false)
+                .With(x => x.CVV_No, "124")
+                .With(x => x.CardNumber, "123")
+              .With(x => x.CardID, 1).Create();
+            var card = _testmapper.Map<Cards>(cardData);
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs:68]: dont hardcode the values

File: BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            //Arrange
+            var fixture = new Fixture();
+
+            var cardData = fixture.Build<CardsVM>().With(a => a.IsDeleted, false).With(x => x.CustomerID, 1)
+                            .With(x => x.CVV_No, "124")
+                            .With(x => x.CardNumber, "123")
+                            .With(x => x.CardID, 1).Create();
+            var card = _testmapper.Map<Cards>(cardData);
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/Integration.Tests/CardRepositoryTest/GetCardDatatableTets.cs:120]: dont hardcode the values

**Review by Backiaraj07 — COMMENTED** [review]
Added comments Kindly fix it

**Review by Backiaraj07 — APPROVED** [review]
PR Verified

**Pull request merged**